### PR TITLE
fix(lib, media-query): support angular/universal

### LINKED ISSUE #187: Using flex-layout with Angular Universal
From my limited knowledge of Node, I gather that anything referencing `Document` or `Window` will throw a server-side error.

That said, I'd like to incorporate flex-layout into my Angular Universal app. Is that possible? My initial attempt (simply following the readme) throws the following error:

> ReferenceError: document is not defined

### PR DESCRIPTION
**Pull request by ThomasBurleson**
Use getDom() for platform-server/universal fixes.
Now gets document object from platform-browser by DI instead of global.

* Add universal application to packages

Fixes #187, #354. Closes #346.

> Thx to @ardatan for PR #346.

### CODE REVIEW

File: src/universal-app/app/splitter/split-handle.directive.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  }
+})
+export class SplitHandleDirective {
+
+  @Output() drag: Observable<{ x: number, y: number }>;
+
+  constructor(ref: ElementRef) {
+    const getMouseEventPosition = (event: MouseEvent) => ({x: event.movementX, y: event.movementY});
+
+    const mouseup$ = Observable.fromEvent(document, 'mouseup');
```
Comment (ardatan) [inline src/universal-app/app/splitter/split-handle.directive.ts]: I think you have to inject 'DOCUMENT' from 'platform-browser'.

Comment (ThomasBurleson) [inline src/universal-app/app/splitter/split-handle.directive.ts]: @ardatan - true. I think even better is to use `ref.nativeElement.ownerDocument`. Your thoughts ?

> BTW, even with injection or `ownerDocument` I am still getting `gulp ci:prerender` errors:
```console
11:14:25] 'prerender' errored after 611 ms
[11:14:25] Error: ERROR TypeError: Invalid event target
    at Function.FromEventObservable.setupSubscription (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/observable/FromEventObservable.js:113:19)
    at FromEventObservable._subscribe (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/observable/FromEventObservable.js:135:29)
    at FromEventObservable.Observable._trySubscribe (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/Observable.js:171:25)
    at FromEventObservable.Observable.subscribe (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/Observable.js:159:65)
    at MapOperator.call (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/operator/map.js:54:23)
    at Observable.subscribe (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/Observable.js:156:22)
    at SwitchMapOperator.call (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/operator/switchMap.js:67:23)
    at Observable.subscribe (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/Observable.js:156:22)
    at SplitDirective.ngAfterContentInit (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/dist/packages/universal-app/app/splitter/split.directive.js:18:41)
    at callProviderLifecycles (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/@angular/core/bundles/core.umd.js:11206:18)
```
Any ideas ?

**Comment by ThomasBurleson** [comment]
@ardatan - BTW, even with injection or `ownerDocument` I am still getting `gulp ci:prerender` errors:

```console
11:14:25] 'prerender' errored after 611 ms
[11:14:25] Error: ERROR TypeError: Invalid event target
    at Function.FromEventObservable.setupSubscription (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/observable/FromEventObservable.js:113:19)
    at FromEventObservable._subscribe (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/observable/FromEventObservable.js:135:29)
    at FromEventObservable.Observable._trySubscribe (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/Observable.js:171:25)
    at FromEventObservable.Observable.subscribe (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/Observable.js:159:65)
    at MapOperator.call (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/operator/map.js:54:23)
    at Observable.subscribe (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/Observable.js:156:22)
    at SwitchMapOperator.call (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/operator/switchMap.js:67:23)
    at Observable.subscribe (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/rxjs/Observable.js:156:22)
    at SplitDirective.ngAfterContentInit (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/dist/packages/universal-app/app/splitter/split.directive.js:18:41)
    at callProviderLifecycles (/Users/thomasburleson/Documents/projects/Google/projects/angular-layouts/node_modules/@angular/core/bundles/core.umd.js:11206:18)
```
Any ideas ?

**Comment by ardatan** [comment]
in 'split-handle-directive.ts', I think it had better to take events into Observable by using HostListener. I changed like that, then ran ci:prerender. It was successful. You can try.
`import {Directive, ElementRef, Output, Inject, EventEmitter, HostListener} from '@angular/core';
import {DOCUMENT} from '@angular/platform-browser';
import {Observable} from 'rxjs/Observable';

import 'rxjs/add/observable/fromEvent';
import 'rxjs/add/operator/takeUntil';
import 'rxjs/add/operator/switchMap';
import 'rxjs/add/operator/map';

@Directive({
  selector: '[ngxSplitHandle]',
  host: {
    class: 'ngx-split-handle',
    title: 'Drag to resize'
  }
})
export class SplitHandleDirective {

  @Output() drag: Observable<{ x: number, y: number }>;
  mousedown$ = new EventEmitter();

  constructor(private ref: ElementRef, @Inject(DOCUMENT) private document: any) { }
  ngAfterViewInit() {
    const getMouseEventPosition = (event: MouseEvent) => ({ x: event.movementX, y: event.movementY });

    const mouseup$ = Observable.fromEvent(this.document, 'mouseup');
    const mousemove$ = Observable.fromEvent(this.document, 'mousemove')
      .map(getMouseEventPosition);
    this.drag = this.mousedown$.switchMap(_ => mousemove$.takeUntil(mouseup$));

  }
  @HostListener('mousedown', ['$event']) onMouseDown(event) {
    this.mousedown$.next(event);
  }

}`

**Comment by ThomasBurleson** [comment]
@ardatan  - fixed it. Need to use `isBrowser()` utility method.

**Comment by ardatan** [comment]
@ThomasBurleson it is ready to merge?

File: src/universal-app/app/splitter/split-handle.directive.ts (hunk 1 above)
Comment (ardatan) [inline src/universal-app/app/splitter/split-handle.directive.ts]: @ThomasBurleson I am not sure if platform-server has `ownerDocument` property implementation.

**Comment by ThomasBurleson** [comment]
@ardatan - we have a rigorous merge process that includes tests with many applications internal to Google. This will be marked as `presubmit`. The merge process may take 2-3 work days.

**Comment by ardatan** [comment]
@ThomasBurleson I understand. Thank you for explanation.
30 Tem 2017 Paz, saat 21:59 tarihinde Thomas Burleson <
notifications@github.com> şunu yazdı:

**Comment by ardatan** [comment]
@ThomasBurleson I have an idea about pre-rendered flex-layout pages. Now when we are using flex-layout with universal, it only applies default styles than changes by current screen.
But here is the thing, we can prevent that bad-looking issue by changing the approach.
1)We can predict screen size by using user-agent from request(which is available in platform-server/not-implemented but could be)
2)We can apply all styles by using media queries.

What do you think?

File: src/lib/flexbox/api/base.ts [hunk 2]
```diff
@@ -117,20 +118,27 @@ export abstract class BaseFxDirective implements OnDestroy, OnChanges {
...
   }
 
   protected _getFlowDirection(target: any, addIfMissing = false): string {
     let value = '';
     if (target) {
-      let directionKeys = Object.keys(applyCssPrefixes({'flex-direction': ''}));
-      let findDirection = (styles) => directionKeys.reduce((direction, key) => {
-        return direction || styles[key];
-      }, null);
+      let immediateValue = getDom().getStyle(target, 'flex-direction');
```
Comment (mmalerba) [inline src/lib/flexbox/api/base.ts]: this logic appears twice, move to helper function?

File: src/demo-app/app/github-issues/splitter/split-handle.directive.ts [hunk 3]
```diff
@@ -19,13 +20,14 @@ export class SplitHandleDirective {
 
   constructor(ref: ElementRef) {
     const getMouseEventPosition = (event: MouseEvent) => ({x: event.movementX, y: event.movementY});
+    const document = ref.nativeElement.ownerDocument;
```
Comment (mmalerba) [inline src/demo-app/app/github-issues/splitter/split-handle.directive.ts]: I see elsewhere you are doing `@Inject(DOCUMENT)` why not same here?

File: src/lib/media-query/match-media.ts [hunk 4]
```diff
@@ -114,9 +116,12 @@ export class MatchMedia {
    * Call window.matchMedia() to build a MediaQueryList; which
    * supports 0..n listeners for activation/deactivation
    */
-  protected  _buildMQL(query: string): MediaQueryList {
-    let canListen = !!(<any>window).matchMedia('all').addListener;
-    return canListen ? (<any>window).matchMedia(query) : <MediaQueryList>{
+  protected _buildMQL(query: string): MediaQueryList {
+    let canListen = (typeof matchMedia != 'undefined');
```
Comment (mmalerba) [inline src/lib/media-query/match-media.ts]: can just do `canListen = !!matchMedia` or not even and just do:

```ts
if (matchMedia) {
  ...
}
return matchMedia ? ...;
```

File: src/universal-app/app/responsive-app.css [hunk 5]
```diff
@@ -0,0 +1,84 @@
+.night-theme {
+  background: #1c2029;
+  color: #cfcfcf;
+}
+
+.handle {
+  outline: none;
+  -webkit-user-select: none;
+  user-select: none;
+  z-index: 9999;
```
Comment (mmalerba) [inline src/universal-app/app/responsive-app.css]: does the z-index need to be so high?

File: src/universal-app/app/responsive-app.css [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  position: relative;
+  line-height: 0;
+}
+
+.handle-row {
+  width: 15px;
+  top: 50%;
+  left: -2px;
+  transform: translateX(-50%) rotate(270deg);
+  cursor: col-resize;;
```
Comment (mmalerba) [inline src/universal-app/app/responsive-app.css]: double `;`

File: src/universal-app/app/responsive-app.css (hunk 5 above)
Comment (ThomasBurleson) [inline src/universal-app/app/responsive-app.css]: I just chose an arbitrary large 100.

Comment (ThomasBurleson) [inline src/universal-app/app/responsive-app.css]: Fixed.

File: src/universal-app/app/responsive-app.css (hunk 6 above)
Comment (ThomasBurleson) [inline src/universal-app/app/responsive-app.css]: Fixed.

File: src/lib/media-query/match-media.ts (hunk 4 above)
Comment (ThomasBurleson) [inline src/lib/media-query/match-media.ts]: I will use the angular Universal `isBrowser()` solution.

File: src/demo-app/app/github-issues/splitter/split-handle.directive.ts (hunk 3 above)
Comment (ThomasBurleson) [inline src/demo-app/app/github-issues/splitter/split-handle.directive.ts]: I will sync the solutions to be the same. Thx.

File: src/lib/flexbox/api/base.ts (hunk 2 above)
Comment (ThomasBurleson) [inline src/lib/flexbox/api/base.ts]: Fixed.

File: src/demo-app/app/github-issues/splitter/split-handle.directive.ts (hunk 3 above)
Comment (ThomasBurleson) [inline src/demo-app/app/github-issues/splitter/split-handle.directive.ts]: Fixed.

**Comment by ThomasBurleson** [comment]
@mmalerba - all comments fixed. Thank you sir.

**Comment by cyrilletuzi** [comment]
With last head build, Universal still does not work with Flex Layout, got these errors : 

```
ReferenceError: document is not defined
    at prepareQueryCSS (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:834:34)
    at MatchMedia.exports.MatchMedia.MatchMedia._buildMQL (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:802:9)
    at MatchMedia.exports.MatchMedia.MatchMedia.registerQuery (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:788:28)
    at /node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:1063:31
    at Array.forEach (native)
    at MediaMonitor.exports.MediaMonitor.MediaMonitor._registerBreakpoints (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:1062:33)
    at new MediaMonitor (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:995:14)
    at MEDIA_MONITOR_PROVIDER_FACTORY (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:1525:29)
    at _callFactory (/node_modules/@angular/core/bundles/core.umd.js:9585:26)
    at _createProviderInstance$1 (/node_modules/@angular/core/bundles/core.umd.js:9521:26)
ERROR { Error: Uncaught (in promise): ReferenceError: window is not defined
ReferenceError: window is not defined
    at MatchMedia.exports.MatchMedia.MatchMedia._buildMQL (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:803:27)
    at MatchMedia.exports.MatchMedia.MatchMedia.registerQuery (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:788:28)
    at /node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:1063:31
    at Array.forEach (native)
    at MediaMonitor.exports.MediaMonitor.MediaMonitor._registerBreakpoints (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:1062:33)
    at new MediaMonitor (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:995:14)
    at MEDIA_MONITOR_PROVIDER_FACTORY (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:1525:29)
    at _callFactory (/node_modules/@angular/core/bundles/core.umd.js:9585:26)
    at _createProviderInstance$1 (/node_modules/@angular/core/bundles/core.umd.js:9521:26)
    at resolveNgModuleDep (/node_modules/@angular/core/bundles/core.umd.js:9503:17)
  
rejection:
   { ReferenceError: window is not defined
       at MatchMedia.exports.MatchMedia.MatchMedia._buildMQL (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:803:27)
       at MatchMedia.exports.MatchMedia.MatchMedia.registerQuery (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:788:28)
       at /node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:1063:31
       at Array.forEach (native)
       at MediaMonitor.exports.MediaMonitor.MediaMonitor._registerBreakpoints (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:1062:33)
       at new MediaMonitor (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:995:14)
       at MEDIA_MONITOR_PROVIDER_FACTORY (/node_modules/@angular/flex-layout/bundles/flex-layout.umd.js:1525:29)
```

Does not crash the server, but stop rendering all the rest of the page.

**Comment by ThomasBurleson** [comment]
@cyrilletuzi - we need your help to reproduce this issue. Do you have a small demo that we can test locally to produce the same error ?

**Comment by cyrilletuzi** [comment]
Sorry, it was just a package synchronisation issue (I updated my app flex layout package but not the server one...), all is OK now and I can close the reflect issue.

Is it possible to release a beta.9 for this ? I suppose if it's not still here yet it's because there are other things coming, but as it's a beta release it is not critical to miss some other points and would be better than rely on nightly builds.

**Comment by ThomasBurleson** [comment]
@cyrilletuzi - thx for the follow-up and confirmation/closure of your issue. We hope to release Beta.9 this week.

**Comment by wesamco** [comment]
great OSS work guys.
happy to hear that material and flex-layout work OOB with Universal now after beta 9 releasi this week 🎉.
@ThomasBurleson what about what @ardatan mentioned above? ;
theres libs for predicting specific device & if Desktop/Mobile from user agent of Express request, we can map it to exact width value for mobile devices and close enough one (1080p?) for desktop, for much better html render.
is there a way to make universal render to specified width?

**Comment by cyrilletuzi** [comment]
@wesamco It's discussed in #373

**Pull request merged**